testbed/sampling/sampling_data_provider.go

```diff
@@ -0,0 +1,155 @@
+	// Stores all the trances that were sampled
+	expectedSampledTraces []ptrace.Traces
+}
+
+type sampledSpanCustomizerFunc func(span ptrace.Span)
+
+type shouldSampleFunc func(traceIDSequence uint64) bool
+
+// SamplingDataProviderOption defines a PollutedDataProvider option.
+type SamplingDataProviderOption func(t *samplingDataProvider)
+
+// WithSampledSpanCustomizer allows you to customize the spans that are sampled.
+func WithSampledSpanCustomizer(customizer sampledSpanCustomizerFunc) SamplingDataProviderOption {
+	return func(tc *samplingDataProvider) {
+		tc.sampledSpanCustomizer = customizer
+	}
+}
+
+// NewSamplingDataProvider creates a new instance of samplingDataProvider
+func NewSamplingDataProvider(opts ...SamplingDataProviderOption) *samplingDataProvider {
+	sdp := samplingDataProvider{
+		// Defaults. These are arbitrary values.
+		// We should configure them through options based on the need
+		spansPerTrace:       6,
+		numTracesToGenerate: 50,
+		shouldSample:        func(traceIDSequence uint64) bool { return traceIDSequence%10 < 3 },
+	}
+
+	for _, opt := range opts {
+		opt(&sdp)
+	}
+
+	return &sdp
+}
+
+// SetLoadGeneratorCounters sets the data providers items generated field to have a ref to the load generators counter
+func (sdp *samplingDataProvider) SetLoadGeneratorCounters(dataItemsGenerated *atomic.Uint64) {
+	sdp.dataItemsGenerated = dataItemsGenerated
+}
+
+// GenerateTraces the boolean in this return reflects whether we are done generating traces or not.
+func (sdp *samplingDataProvider) GenerateTraces() (ptrace.Traces, bool) {
+	if sdp.tracesGenerated >= sdp.numTracesToGenerate {
+		return ptrace.NewTraces(), true
+	}
+	sdp.tracesGenerated += 1
+
+	traceData := sdp.generateTrace()
+
+	return traceData, false
+
+}
+
+func (sdp *samplingDataProvider) sampledSpansGenerated() uint64 {
+	return sdp.sampledSpansCounter.Load()
+}
+
+func (sdp *samplingDataProvider) generateTrace() ptrace.Traces {
+	traceData := ptrace.NewTraces()
+
+	spans := traceData.ResourceSpans().AppendEmpty().ScopeSpans().AppendEmpty().Spans()
+	spans.EnsureCapacity(sdp.spansPerTrace)
+
+	traceID := sdp.traceIDSequence.Add(1)
+	shouldSample := sdp.shouldSample(traceID)
+	// TODO: should we be using ItemsPerBatch to determine how many spans are in a trace?
+	for i := 0; i < sdp.spansPerTrace; i++ {
+
+		startTime := time.Now()
+		endTime := startTime.Add(time.Millisecond)
+
+		spanID := sdp.dataItemsGenerated.Add(1)
+
+		span := spans.AppendEmpty()
+
+		// Create a span.
+		span.SetTraceID(uInt64ToTraceID(0, traceID))
+		span.SetSpanID(uInt64ToSpanID(spanID))
+		span.SetKind(ptrace.SpanKindClient)
+		attrs := span.Attributes()
+
+		attrs.PutInt("some.attribute", int64(traceID))
+
+		span.SetStartTimestamp(pcommon.NewTimestampFromTime(startTime))
+		span.SetEndTimestamp(pcommon.NewTimestampFromTime(endTime))
+
+		if shouldSample {
+			sdp.sampledSpansCounter.Add(1)
+			sdp.sampledSpanCustomizer(span)
+		}
```

Comment:
All the spans of a trace will have the customization.

We can still customize the full trace duration by keeping track of the state in the customizing funcation.

But that is a good call for the future. Should we customize the full trace instead of customize only individual spans?